### PR TITLE
[Merged by Bors] - feat: port InformationTheory.Hamming

### PR DESCRIPTION
**Pull request by urkud**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/InformationTheory/Hamming.lean [hunk 1]
```diff
@@ -425,51 +416,46 @@ theorem dist_eq_hammingDist (x y : Hamming β) :
...
     nndist x y = hammingDist (ofHamming x) (ofHamming y) :=
   rfl
 #align hamming.nndist_eq_hamming_dist Hamming.nndist_eq_hammingDist
 
-instance : MetricSpace (Hamming β) :=
-  { Hamming.pseudoMetricSpace with
-    eq_of_dist_eq_zero := by
-      push_cast
-      exact_mod_cast @eq_of_hammingDist_eq_zero _ _ _ _ }
+instance : DiscreteTopology (Hamming β) := ⟨rfl⟩
```
Comment (eric-wieser) [inline Mathlib/InformationTheory/Hamming.lean:456]: Where did this come from?
```suggestion
-- porting note: new
instance : DiscreteTopology (Hamming β) := ⟨rfl⟩
```

File: Mathlib/InformationTheory/Hamming.lean [hunk 2]
```diff
@@ -479,22 +465,15 @@ theorem norm_eq_hammingNorm [∀ i, Zero (β i)] (x : Hamming β) : ‖x‖ = ha
   rfl
 #align hamming.norm_eq_hamming_norm Hamming.norm_eq_hammingNorm
 
-instance [∀ i, AddCommGroup (β i)] : SeminormedAddCommGroup (Hamming β) :=
-  { Pi.addCommGroup with
-    dist_eq := by
-      push_cast
-      exact_mod_cast hammingDist_eq_hammingNorm }
+instance [∀ i, AddCommGroup (β i)] : NormedAddCommGroup (Hamming β) where
```
Comment (eric-wieser) [inline Mathlib/InformationTheory/Hamming.lean:468]:
```suggestion
-- porting note: merged `SeminormedAddCommGroup` and `NormedAddCommGroup` instances

instance [∀ i, AddCommGroup (β i)] : NormedAddCommGroup (Hamming β) where
```

File: Mathlib/InformationTheory/Hamming.lean [hunk 3]
```diff
@@ -335,12 +335,12 @@ theorem ofHamming_toHamming (x : ∀ i, β i) : ofHamming (toHamming x) = x :=
   rfl
 #align hamming.of_hamming_to_hamming Hamming.ofHamming_toHamming
 
--- @[simp] -- Porting note: simp can prove this
+@[simp, nolint simpNF] -- useful for dsimp
```
Comment (urkud) [inline Mathlib/InformationTheory/Hamming.lean:338]: @eric-wieser Can `dsimp` use `Iff.rfl` lemmas in Lean 4?

**Pull request closed without merging**